### PR TITLE
[WIP] Added AbstractWrappedCollection support.

### PR DESCRIPTION
**Pull request by guilhermeblanco**

### CODE REVIEW

File: lib/Doctrine/Common/Collections/AbstractWrappedCollection.php [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+use Closure;
+
+/**
+ * An AbstractWrappedCollection is a Collection implementation that wraps another Collection.
+ *
+ * @since  1.3
+ * @author Guilherme Blanco <guilhermeblanco@hotmail.com>
+ */
+abstract class AbstractWrappedCollection extends ArrayCollection
```
Comment (deeky666) [inline lib/Doctrine/Common/Collections/AbstractWrappedCollection.php]: Why extending `ArrayCollection`? Isn't implementing `Collection` interface directly a better choice? No need to add a dependency on `ArrayCollection` here IMO as your contructor isn't tied to `ArrayCollection` either.

File: lib/Doctrine/Common/Collections/AbstractWrappedCollection.php [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    protected $collection;
+
+    /**
+     * Initializes a new ArrayCollection.
+     *
+     * @param array $elements
+     */
+    public function __construct(Collection $collection = null)
+    {
+        $this->collection = $collection ?: new ArrayCollection();
```
Comment (deeky666) [inline lib/Doctrine/Common/Collections/AbstractWrappedCollection.php]: I think I would not make a default instance here, that is too much implementation detail of the abstract collection wrapper IMO and should rather be a decision to be made in the concrete implementation.
I would suggest making `$collection` a required parameter instead.

File: lib/Doctrine/Common/Collections/AbstractWrappedCollection.php [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+{
+    /**
+     * @var Collection
+     */
+    protected $collection;
+
+    /**
+     * Initializes a new ArrayCollection.
+     *
+     * @param array $elements
```
Comment (deeky666) [inline lib/Doctrine/Common/Collections/AbstractWrappedCollection.php]: Please fix the DocBlock param and description here.

**Comment by Ocramius** [comment]
The naming does not express the intent here: why is this needed? Where is the change of behavior, compared to a normal collection?

**Comment by beberlei** [comment]
Again here: Why? You need to explain why we need this :)

File: lib/Doctrine/Common/Collections/AbstractWrappedCollection.php (hunk 1 above)
Comment (guilhermeblanco) [inline lib/Doctrine/Common/Collections/AbstractWrappedCollection.php]: Makes sense... will address this

**Comment by guilhermeblanco** [comment]
@Ocramius @beberlei It's a Collection that wraps another one.
Right now, PersistentCollection already does that by wrapping a Collection of elements.
I'm implementing this to:
# Ditch lines of code from PersistentCollection in ORM
# Add support for custom collections which is currently WIP here: https://github.com/doctrine/doctrine2/compare/custom-collections

File: lib/Doctrine/Common/Collections/AbstractWrappedCollection.php (hunk 2 above)
Comment (guilhermeblanco) [inline lib/Doctrine/Common/Collections/AbstractWrappedCollection.php]: I can't... here's the final idea:

```php
/**
 * @ORM\Entity
 */
class User
{
    private $roles;

    public function __construct()
    {
        $this->roles = new RoleCollection();
    }
}
```

**Comment by danizord** [comment]
I think `AbstractCollectionWrapper` is a more clear naming

**Comment by stof** [comment]
@guilhermeblanco how does this compare to the current AbstractLazyCollection ?

**Comment by guilhermeblanco** [comment]
@stof There're 2 main differences:
* No initialization
* `partition()` must return instances of static

**Comment by stof** [comment]
@guilhermeblanco a PersistentCollection is actually a lazy one currently. It has initialization logic

**Comment by guilhermeblanco** [comment]
@stof I realized that after you pointed out... will update on ORM branch later today.
It still does not invalidate the purpose of this class.

**Comment by Ocramius** [comment]
I don't see a correlation between `AbstractCollectionWrapper` and the functionality being provided though: really needs something more explicit/clear.

**Comment by danizord** [comment]
Well, this is not wrapped, but it wraps.

**Comment by Ocramius** [comment]
> it wraps.

yes, but what does it **do**?

**Comment by Ocramius** [comment]
Closing here - no longer needed.

**Pull request closed without merging**